### PR TITLE
GraphQL: User Settings & Registration

### PR DESCRIPTION
**Pull request by Jonasdoubleyou**
This adds the following common implementations:
- Decisions (a boolean and a reason), to communicate reasons why a certain action is not allowed to the frontend, so that the frontend can show a proper warning to the user. That way we do not ave to replicate backend logic in the frontend
- `logInContext` for logging within GraphQL. With this one can log consistently with a unique session id (without leaking the whole session id (= the user's secret))
- Elevated Mode: `Role.ADMIN` and `Role.SCREENER` can call certain mutations on behalf of another user, passing in a studentId or pupilId respectively (this is centrally implemented through `getSession(Pupil|Student)`
- Login & Role assignment are split up, so that roles of a session can be reassigned (e.g. due to registration / settings change) 
- Logging in to GraphQL via authToken (which was sent to the user via E-Mail) also sets `verifiedAt` and `verification` properly. In the REST login, this was only done on dedicated endpoint, which lead to some users ignoring the verification E-Mail. This should make the verification E-Mail obsolete long-term (as users indirectly verify the E-Mail through a log in)
- Students and Pupils get assigned more roles (`TUTOR`, `TUTEE`, `PROJECT_COACH`, `PROJECT_COACHEE`, `INSTRUCTOR`, `PARTICIPANT`), if they (a) have their E-Mail verified and (b) are properly screened. That way screening is validated at one central place
- `@RateLimit` annotation to rate limit a certain query / mutation per IP address. This should prevent basic attacks to find user E-Mails and the alike. 
- The `getScreeningURL` method was removed from the TypeORM entity, as we want to move to Prisma anyways and the method was not user dependent anyways (returned a constant string)

The following Queries are added:
- `schools` to everyone, to be able to query the list of schools
- `students` to Admins

The following fields are added:
- `pupil { gradeAsInt }` to get away from the "grade string"
- `pupil { canRequestMatch { allowed reason } }` to check whether the pupil can request a match, and if not provide a reason why
- `student { canRequestMatch { allowed reason } }` analogously for students

The following mutations are changed:
- `matchDissolve` is also released to Owners (the student and the pupil involved in the match)
- `pupilCreateMatchRequest` is now also released to tutees
- `pupilDeleteMatchRequest` is now also released to tutees
- `studentCreateMatchRequest` is now also released to tutors
- `studentDeleteMatchRequest` is now also released to tutors

The following mutations are added:
- `meRegisterStudent` to register a new student account (this is now shared for Instructors, Project Coaches etc. and only initializes the basic entity with the firstname, lastname etc.)
- `meRegisterPupil` to register a new pupil account (this is now shared for Tutees, Coachees, etc. and only intializes the basic entity with the firstname, lastname, etc.)
- `meUpdate` to update the basic user data
- `meDeactivate` to deactivate the user account
- `meActivate`  to reactivate the user account
- `meBecomeInstructor` to change the student account to an instructor account
- `meBecomeTutor` to change the student account to a tutor account
- `meBecomeProjectCoach` to change the student account to a project coach

- `meBecomeTutee` to change the pupil account into a tutee account
- `meBecomeProjectCoachee` to change the pupil account into a project coachee account
- `meBecomeStatePupil` to change the pupil account into a state pupil account
- `isEmailAvailable` to check whether an e-mail is already taken (unauthenticated access, rate limited)

### CODE REVIEW

**Review by realmayus — COMMENTED** [review]
LGTM!

File: common/pupil/registration.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+            // the authToken is used to verify the e-mail instead
+            verification: uuidv4()
+        }
+    });
+
+    // TODO: Create a new E-Mail for registration
+    // TODO: Send auth token with this
+    await Notification.actionTaken(pupil, "pupil_registration_started", { redirectTo: data.redirectTo });
+    await logTransaction("verificationRequets", pupil, {});
```
[the remark this answers is not in the document]
Comment (Jonasdoubleyou) [reply common/pupil/registration.ts:85]: True, in fact this typo is already present in the current transaction logging. So keeping it is backwards compatible, though maybe this is not needed as the transaction log is used sporadically anyways. Any opinion on this @lsc0498 ? (And good catch by the way ;))

File: common/student/registration.ts [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+export async function becomeProjectCoach(student: Student, data: BecomeProjectCoachData) {
+    if (student.isProjectCoach) {
+        throw new Error(`Student is already a coach`);
+    }
+
+    const { hasJufoCertificate, isUniversityStudent, jufoPastParticipationInfo, projectFields, wasJufoParticipant } = data;
+
+    if (projectFields) {
```
[the remark this answers is not in the document]
Comment (Jonasdoubleyou) [reply common/student/registration.ts:156]: True, moved to `setProjectFields` function and added an explanatory comment.

**Review by realmayus — APPROVED** [review]

File: common/pupil/registration.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (lsc0498) [reply common/pupil/registration.ts:85]: Yes good point. It might be misleading if there are two variants in the data.

**Review by lsc0498 — APPROVED** [review]
Pretty good job!🎉

**Pull request merged**